### PR TITLE
Stitching Upload Components to support multiple Upload Strategies

### PR DESCRIPTION
**Pull request by anchita-g**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

One of the PRs to fix https://github.com/google/android-fhir/issues/760

**Description**
We connect the four components:` LocalChangeFetcher`, `PatchGenerator` , `RequestGenerator` and `ResourceConsolidator`  for upload sync to work according to any type of `UploadStrategy` provided. 

Since we have not implemented al the types of the components, with the implementations we have so far, we can only use a few `UploadStrategy`. 
The connection of the components is mainly established by maintaining a mapping of the `LocalChanges` to the respective outputs in each stage. For example, `PatchGenerator` is expected to output `Patch` as well as the respective list of `LocalChange` from which each `Patch` was generated. 

For `RequestGenerator`, using the mapped `Patch` for `LocalChanges`, the request generator should output the `UploadRequest` generated as well as the `LocalChanges` of the `Patches` from which this request was generated. 
For `BundleRequestGenerator`, the mapping to the `LocalChanges` is at the level of each entry in the `BundleRequest`. 

This would help in the `ResourceConsolidator` where in each `BundleEntry` is like an independent request being executed with its own corresponding response, hence, we would need to know the `LocalChanges` which result in that particular BundleEntry request.

**Alternative(s) considered**
NA

**Type**
Choose one: Feature
**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/sync/upload/patch/PatchGenerator.kt [hunk 1]
```diff
@@ -54,3 +55,13 @@ internal sealed class PatchGeneratorMode {
 
   object PerChange : PatchGeneratorMode()
 }
+
+/**
+ * Structure to maintain the mapping between [List<[LocalChange]>] and the [Patch] generated from
+ * those changes. This class should be used by any implementation of [PatchGenerator] to output the
+ * [Patch] in this format.
+ */
+internal data class PatchGeneratorOutput(
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/patch/PatchGenerator.kt:64]: I feel we are over-complicating the code by introducing this mapping. See suggestion in Uploader for how we might be able to make this simpler

File: engine/src/main/java/com/google/android/fhir/sync/upload/patch/PatchGenerator.kt [hunk 2]
```diff
@@ -33,7 +34,7 @@ internal interface PatchGenerator {
    * NOTE: different implementations may have requirements on the size of [localChanges] and output
    * certain numbers of [Patch]es.
    */
-  fun generate(localChanges: List<LocalChange>): List<Patch>
+  fun generate(localChanges: List<LocalChange>): List<PatchGeneratorOutput>
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/patch/PatchGenerator.kt:37]: Revert to how it was

File: engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt [hunk 3]
```diff
@@ -38,67 +43,131 @@ import timber.log.Timber
...
-
+internal class Uploader(
+  private val dataSource: DataSource,
+  private val patchGenerator: PatchGenerator,
+  private val requestGenerator: UploadRequestGenerator,
+) {
   suspend fun upload(localChanges: List<LocalChange>): UploadSyncResult {
     val patches = patchGenerator.generate(localChanges)
-    val requests = requestGenerator.generateUploadRequests(patches)
+    val mappedRequests = requestGenerator.generateUploadRequests(patches)
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt:53]: What do you think about doing the following to simplify:

* Revert `patchGenerator.generate(localChanges)` to return a `List<Patch>`
*  Revert `generateUploadRequests` back to `generateUploadRequests(patches: List<Patch>): List<UploadRequest>`
*  Let's do the chunking of the transaction bundles here in the uploader
```suggestion
 val mappedRequests = when (requestGenerator) {
      is UrlRequestGenerator -> requestGenerator.generateUploadRequests(patches)
      is TransactionBundleGenerator -> {
        patches.chunked(requestGenerator.generatedBundleSize)
          .flatMap { requestGenerator.generateUploadRequests(it) }
      }
      else -> error("")
    }
```


These changes I feel will allow us to simplify and get rid of the `PatchGeneratorOutput` class. wdyt?

File: engine/src/main/java/com/google/android/fhir/sync/upload/UploadStrategy.kt [hunk 4]
```diff
@@ -35,33 +34,33 @@ private constructor(
...
-      LocalChangesFetchMode.PerResource,
-      PatchGeneratorMode.PerResource,
-      UploadRequestGeneratorMode.UrlRequest(HttpVerb.POST, HttpVerb.PATCH),
-    )
+  //  object SingleResourcePost :
+  //    UploadStrategy(
+  //      LocalChangesFetchMode.PerResource,
+  //      PatchGeneratorMode.PerResource,
+  //      UploadRequestGeneratorMode.UrlRequest(HttpVerb.POST, HttpVerb.PATCH),
+  //    )
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/upload/UploadStrategy.kt]: Uncommented the UploadStrategies , made them private and moved them to a not implemented section.

File: engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt (hunk 3 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt:53]: I think these changes are fine as they hep us keep our uploader class clean.

File: engine/src/main/java/com/google/android/fhir/sync/upload/patch/PatchGenerator.kt (hunk 2 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/upload/patch/PatchGenerator.kt:37]: Keeping it for now as the Local Changes associated with the Patch might be required in some cases during resolution after the successful upload.

File: engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGenerator.kt [hunk 5]
```diff
@@ -53,3 +62,17 @@ internal object UploadRequestGeneratorFactory {
         )
     }
 }
+
+internal sealed class UploadRequestGeneratorOutput(
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGenerator.kt:66]: similar to patch, should we name this to UploadRequestMapping, as well as the inheriting data classes?
```suggestion
internal sealed class UploadRequestGeneratorOutput(
```

File: engine/src/main/java/com/google/android/fhir/testing/Utilities.kt [hunk 6]
```diff
@@ -222,3 +228,14 @@ class BundleDataSource(val onPostBundle: suspend (Bundle) -> Resource) : DataSou
   override suspend fun upload(request: UploadRequest) =
     onPostBundle((request as BundleUploadRequest).resource)
 }
+
+class UrlRequestDataSource(val onUrlRequestSend: suspend (UrlUploadRequest) -> Resource) :
+  DataSource {
+
+  override suspend fun download(downloadRequest: DownloadRequest): Resource {
+    TODO("Not yet implemented")
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/testing/Utilities.kt:236]: ?

File: engine/src/main/java/com/google/android/fhir/sync/upload/request/UrlRequestGenerator.kt [hunk 7]
```diff
@@ -29,8 +30,15 @@ internal class UrlRequestGenerator(
   private val getUrlRequestForPatch: (patch: Patch) -> UrlUploadRequest,
 ) : UploadRequestGenerator {
 
-  override fun generateUploadRequests(patches: List<Patch>): List<UrlUploadRequest> =
-    patches.map { getUrlRequestForPatch(it) }
+  override fun generateUploadRequests(
+    patches: List<PatchGeneratorOutput>,
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/request/UrlRequestGenerator.kt:34]:
```suggestion
    mappedPatches: List<PatchGeneratorOutput>,
```

File: engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt [hunk 8]
```diff
@@ -38,67 +43,131 @@ import timber.log.Timber
...
-    val localChanges: List<LocalChange>,
-    val responseResources: List<Resource>,
+    val uploadResponses: List<SuccessfulUploadRequestResponse>,
   ) : UploadSyncResult()
 
   data class Failure(val syncError: ResourceSyncException, val localChangeToken: LocalChangeToken) :
     UploadSyncResult()
 }
+
+sealed class SuccessfulUploadRequestResponse(
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt:160]: rename to be more accurate. here and inherited classes below
```suggestion
sealed class SuccessfulUploadResponseMapping(
```

File: engine/src/main/java/com/google/android/fhir/sync/upload/patch/PatchGenerator.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/upload/patch/PatchGenerator.kt]: Done

File: engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt [hunk 9]
```diff
@@ -38,67 +43,131 @@ import timber.log.Timber
...
-  private val patchGenerator = PerResourcePatchGenerator
-  private val requestGenerator = TransactionBundleGenerator.getDefault()
-
+internal class Uploader(
+  private val dataSource: DataSource,
+  private val patchGenerator: PatchGenerator,
+  private val requestGenerator: UploadRequestGenerator,
+) {
   suspend fun upload(localChanges: List<LocalChange>): UploadSyncResult {
     val patches = patchGenerator.generate(localChanges)
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt]: Done

File: engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGenerator.kt (hunk 5 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGenerator.kt]: Done

File: engine/src/main/java/com/google/android/fhir/testing/Utilities.kt (hunk 6 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/testing/Utilities.kt:236]: Its only used for upload testing I guess. That's why the download functionality hasn't been implemented in this test data source.

File: engine/src/main/java/com/google/android/fhir/sync/upload/request/UrlRequestGenerator.kt (hunk 7 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/upload/request/UrlRequestGenerator.kt]: Done

File: engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt (hunk 8 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt]: Done

**Review by omarismail94 — APPROVED** [review]

**Pull request merged**